keps/115-Subgroup-support/README.md

```diff
@@ -0,0 +1,223 @@
+
+The overall workflow will look like this 
+
+![kep-115](https://github.com/kubernetes-sigs/lws/assets/86417275/ff9fc93d-c738-4c09-abc8-50a7b16d49df)
+
+Suppose we have an LWS deployment with 2 replicas, subGroupSize 4, size 8, and the following annotations: 
+- `leaderworkerset.sigs.k8s.io/exclusive-topology: cloud.google.com/gke-placement-group` 
+- `leaderworkerset.sigs.k8s.io/subgroup-exclusive-topology: cloud.google.com/gke-nodepool`
+
+The leader pod will first be scheduled on a placement group. Once a worker pod is created, it will follow the leader to the placement group. 
+Afterward, it will be scheduled into a nodepool that has other pods with the same subgroup index. So, the placement will look something like this
+
+![kep-115 scheduling](https://github.com/kubernetes-sigs/lws/assets/86417275/8127f005-2f50-4648-aa1a-6b37cb3b813a)
+
+### Implementation
+- Two new annotations will be added
+  - `leaderworkerset.sigs.k8s.io/subgroup-exclusive-topology`
+  - `leaderworkerset.sigs.k8s.io/subgroup-size` 
+- Three new labels will be added,
+  - `leaderworkerset.sigs.k8s.io/subgroup-index = worker-index/subGroupSize`
+    - Tracks which subgroup the pod is part of 
+  - `leaderworkerset.sigs.k8s.io/subgroup-worker-index = worker-index%subGroupSize`
+    - index/identity of a pod inside the pod's subgroup
+  - `leaderworkerset.sigs.k8s.io/subgroup-key` 
+    - Pods that are part of the same subgroup will have an annotation that is a unique hash value will be generated from the name of the leader, and the subgroup-index
+
+To support exclusive placement at the subgroup level, the pod webhook will inject the new labels, and set the pod affinity/anti-affinity on all the pods. If both levels of pod affinity are set, then the leader pod will contain two pod affinities, while the workers will have a node selector, and a single pod affinity set. 
+
+### TPU Environment Variable Injection
+Because the value of the TPU environmental variables will vary between subgroups, the way they are injected will be extended. `TPU_WORKER_ID` will be the value of `subgroup-worker-index`, while the value of `TPU_WORKER_HOSTNAMES` will only be a list of the pods in the same subgroup.
+
+LeaderWorkerSet supports the leader not requesting TPU resources. This raises a problem when determining the values of `subgroup-index` and `subgroup-worker-index`. Suppose we have two TPU slices with two hosts each. Since the leader doesn’t request TPU resources, there will be four workers + the leader, meaning that one of the workers will have a worker index of four. Because of the way the new labels are calculated, this worker will have a subgroup-index of two, creating three subgroup indices (0,1,2) even though there are only two TPU slices.
```

Comment:
I can somehow know what you want to do here, but can't fully understand the words, like `we have two TPU slices with two hosts each`, maybe an example would be great or something else to make this more clear, this will benefit the future readers a lot. Thanks.

Reply:
I'll add a graph for this as well

Reply:
So the subgroupsize here is 2? But why the worker index 4 has a subgroup-worker-index = 3, doesn't it be (workerIndex - 1) % subGroupSize = (4-1) % 2 = 1.

Reply:
made a mistake, fix it

Reply:
You mentioned `meaning that one of the workers will have a worker index of four,` I think you mean the `Worker 4` in the pic, then `Because of the way the new labels are calculated, this worker will have a subgroup-index of two`, according to the formula, `leaderworkerset.sigs.k8s.io/subgroup-index = (workerIndex - 1) / subGroupSize = (4-1) / 2 =1`, but you mentioned the result is two, anything I misunderstood here?

Reply:
The written example is to show the problem that arises if normal behavior is followed, while the diagram is the fixed version.